rpgpy/spcutil.py

```diff
@@ -0,0 +1,150 @@
+from typing import Dict, List, Tuple
+
+import numpy as np
+from numba import jit
+
+
+def spectra2moments(LV0data: Dict, LV0meta: Dict, spec_var: str='TotSpec'):
+    """
+    This routine calculates the radar moments: reflectivity, mean Doppler velocity, spectrum width, skewness and
+    kurtosis from compressed level 0 spectrum files (NoiseFactor > 0) of the 94 GHz RPG cloud radar.
+
+    Args:
+        LV0data: lv0 nD variables
+        LV0meta: lv0 meta data
+
+    Return:
+        container_dict: dictionary of larda containers, including larda container for Ze, VEL, sw, skew, kurt
+
+    """
+    # initialize variables:
+    n_ts, n_rg = LV0data[spec_var].shape[:2]
+    moments = np.full((n_ts, n_rg, 5), np.nan)
+
+    spec_lin = LV0data[spec_var].copy()
+    mask = spec_lin <= 0.0
+    spec_lin[mask] = 0.0
+
+    # combine the mask for "contains signal" with "signal has more than 1 spectral line"
+    mask2D = np.all(mask, axis=2)
+    ranges = np.append(LV0meta['RngOffs'], LV0meta['RAltN'])
+
+    for iC in range(LV0meta['SequN']):
+        Dopp_res = np.mean(np.diff(LV0meta['velocity_vectors'][iC]))
+
+        for iR in range(ranges[iC], ranges[iC + 1]):
+            for iT in range(n_ts):
+                if mask2D[iT, iR]: continue
+                _, (lb, rb) = find_peak_edges(spec_lin[iT, iR, :])
+                moments[iT, iR, :] = radar_moment_calculation(
+                    spec_lin[iT, iR, lb:rb],
+                    LV0meta['velocity_vectors'][iC][lb:rb]
+                )
+
+        # shift mean Doppler velocity by half a bin
+        moments[:, ranges[iC]:ranges[iC + 1], 1] -= Dopp_res / 2.0
+
+    # create the mask where invalid values have been encountered
+    invalid_mask = np.full((n_ts, n_rg), True)
+    invalid_mask[np.where(moments[:, :, 0] > 0.0)] = False
+
+    moments = {var: moments[:, :, i] for i, var in enumerate(['Ze', 'MeanVel', 'SpecWidth', 'Skewn', 'Kurt'])}
+
+    for mom in moments.keys():
+        moments[mom][invalid_mask] = -999.0
+
+    return moments
+
+
+@jit(nopython=True, fastmath=True)
+def radar_moment_calculation(signal: np.array, vel_bins: np.array) -> np.array:
+    """
+    Calculation of radar moments: reflectivity, mean Doppler velocity, spectral width,
+        skewness, and kurtosis of one Doppler spectrum. Optimized for the use of Numba.
+
+    Note:
+        Divide the signal_sum by 2 because vertical and horizontal channel are added.
+        Subtract half of of the Doppler resolution from mean Doppler velocity, because
```

Comment:
because... ?

Reply:
This is based on an email conversation with the radar software developer. Tests showed the division by two is necessary to match the reflectivity values in the LV1. Latest tests showed that this differs for newer RPG radar software versions.